src/main/java/frc/trigon/robot/poseestimation/robotposesources/RobotPoseSource.java

```diff
@@ -92,13 +126,29 @@ private boolean isNewTimestamp() {
         return true;
     }
 
-    private Pose3d doubleArrayToPose3d(double[] doubleArray) {
-        if (doubleArray == null || doubleArray.length != 6)
-            return null;
+    private void logVisibleTags() {
+        if (inputs.hasResult) {
+            Logger.recordOutput("VisibleTags/" + this.getName(), new Pose2d[0]);
+            return;
+        }
 
-        return new Pose3d(
-                new Translation3d(doubleArray[0], doubleArray[1], doubleArray[2]),
-                new Rotation3d(doubleArray[3], doubleArray[4], doubleArray[5])
-        );
+        final Pose2d[] visibleTagPoses = new Pose2d[inputs.visibleTagIDs.length];
+        for (int i = 0; i < visibleTagPoses.length; i++)
+            visibleTagPoses[i] = FieldConstants.TAG_ID_TO_POSE.get(i).toPose2d();
+        Logger.recordOutput("VisibleTags/" + this.getName(), visibleTagPoses);
+    }
+
+    private Matrix<N3, N1> solvePNPAverageDistanceToStdDevs(double translationsStdExponent, double thetaStdExponent, int visibleTags) {
+        final double translationStd = translationsStdExponent * Math.pow(inputs.averageDistanceFromAllTags, 2) / (visibleTags * visibleTags);
+        final double thetaStd = thetaStdExponent * Math.pow(inputs.averageDistanceFromAllTags, 2) / visibleTags;
+
+        return VecBuilder.fill(translationStd, translationStd, thetaStd);
+    }
+
+    private Matrix<N3, N1> assumedHeadingAverageDistanceToStdDevs(double assumedRobotHeadingPose, int visibleTags) {
```

Comment:
These values should be from the contructor (except visibleTags, which should always be one btw, so no need for it)
I assume `assumedRobotHeadingPose` is `translationStdExponent`